File: Lab 3/Lab_3.py
def SJN():
    jobNum = int(input("Enter the number of job(s): "))
    job = [None]*jobNum
    for x in range(jobNum):
        aUnicode = ord("A")+x
        job[x] = int(input("Enter job {} CPU cycle: ".format(chr(aUnicode))))
        aUnicode += 1
    job.sort()
    total = totalSJN(job, jobNum)
    print("Average turnaround time: {}".format(total))

def totalSJN(job, jobNum):
    total = 0
    for x in range(jobNum):
        total += totalRecurse(job, x)
    avgTotal = total/jobNum
    return avgTotal

def totalRecurse(job, x):
    if x > 0:
        total = job[x] + totalRecurse(job, x-1)
    else:
        total = job[0]
    return total

def RR():
    jobNum = int(input("Enter the number of job(s): "))
    job = [None]*jobNum
    for x in range(jobNum):
        aUnicode = ord("A")+x
        job[x] = int(input("Enter job {} CPU cycle: ".format(chr(aUnicode))))
        aUnicode += 1
    jobArrive = [None]*jobNum
    for x in range(jobNum):
        aUnicode = ord("A")+x
        jobArrive[x] = int(input("Enter job {} arrival time: ".format(chr(aUnicode))))
        aUnicode += 1
    timeQuantum = int(input("Enter time quantum: "))
    jobFinish = [0]*jobNum
    jobCheck = [False]*jobNum
    loopCheck = True
    timeQuantumAccumulate = 0
    while not all(jobCheck):
        for x in range(jobNum):
            if job[x] >= timeQuantum and jobCheck[x] == False:
                job[x] -= timeQuantum
                timeQuantumAccumulate += timeQuantum
                hold = timeQuantumAccumulate
                jobFinish[x] = hold
            elif job[x] < timeQuantum and jobCheck[x] == False:
                timeQuantumAccumulate += job[x]
                job[x] = 0
                hold = timeQuantumAccumulate
                jobFinish[x] = hold
            if job[x] == 0 and jobCheck[x] == False:
                    jobCheck[x] = True
    total = 0
    for x in range(jobNum):
        total += jobFinish[x] - jobArrive[x]
    print("Average turnaround time: {}".format(total/jobNum))
    
choice = int(input("Enter 1 for SJN or 2 for Round Robin: "))
if choice == 1:
    SJN()
elif choice == 2:
    RR()
else:
    print("Invalid Input")

File: Lab 3/Lab_3.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

void SJN();
int cmpfunc(const void * a, const void * b);
float totalSJN(int job[], int jobNum);
int totalRecurse(int job[], int x);
void RR();
bool loopCheck(bool jobCheck[], int jobNum);

int main(){
    printf("Enter 1 for SJN or 2 for Round Robin: ");
    int choice;
    scanf("%d", &choice);
    if(choice == 1)
        SJN();
    else if(choice == 2)
        RR();
    else
     printf("Invalid Input");
}


void SJN(){
    printf("Enter the number of job(s): ");
    int jobNum;
    scanf("%d", &jobNum);
    int job[jobNum];
    for(int i=0; i<jobNum; i++){
        printf("Enter job %c CPU cycle: ",i+65);
        scanf("%d", &job[i]);
    }
    qsort(job, jobNum, sizeof(int), cmpfunc);
    float total = totalSJN(job, jobNum);
    printf("Average turnaround time: %f", total);
}

int cmpfunc (const void * a, const void * b) {
   return ( *(int*)a - *(int*)b );
}

float totalSJN(int job[], int jobNum){
    int total = 0;
    for(int i=0; i<jobNum; i++){
        total += totalRecurse(job, i);
    }
    float avgtotal = total/(float)jobNum;
    return avgtotal;
}

int totalRecurse(int job[], int x){
    int total = 0;
    if(x > 0)
        total = job[x] + totalRecurse(job, x-1);
    else
        total = job[0];
    return total;
}

void RR(){
    printf("Enter the number of job(s): ");
    int jobNum;
    scanf("%d", &jobNum);
    int job[jobNum];
    for(int i=0; i<jobNum; i++){
        printf("Enter job %c CPU cycle: ",i+65);
        scanf("%d", &job[i]);
    }
    int jobArrive[jobNum];
    for(int i=0; i<jobNum; i++){
        printf("Enter job %c arrival time: ",i+65);
        scanf("%d", &jobArrive[i]);
    }
    int timeQuantum;
    printf("Enter time quantum: ");
    scanf("%d", &timeQuantum);
    int jobFinish[jobNum];
    bool jobCheck[jobNum];
    for(int i=0; i<jobNum; i++)
        jobCheck[i] = false;
    int timeQuantumAccumulate = 0;
    while (loopCheck(jobCheck, jobNum)){
        for(int i=0; i<jobNum; i++){
            if((job[i] >= timeQuantum) && (jobCheck[i] == false)){
                job[i] -= timeQuantum;
                timeQuantumAccumulate += timeQuantum;
                int hold = timeQuantumAccumulate;
                jobFinish[i] = hold;
            }
            else if((job[i] < timeQuantum) && (jobCheck[i] == false)){
                timeQuantumAccumulate += job[i];
                job[i] = 0;
                int hold = timeQuantumAccumulate;
                jobFinish[i] = hold;
            }
            if((job[i] == 0) && (jobCheck[i] == false))
                jobCheck[i] = true;
        }
    }
    int total = 0;
    for(int i=0; i<jobNum; i++)
        total += jobFinish[i] - jobArrive[i];
    float avg;
    avg = total/(float)jobNum;
    printf("Average turnaround time: %f", avg);
}

bool loopCheck(bool jobCheck[], int jobNum){
    int check = 0;
    for(int i=0; i<jobNum; i++){
        if(jobCheck[i] == true)
            check++;
    }
    if(check == jobNum)
        return false;
    else
        return true;
}